File: JS/No_Framework/Factory/js/concrete.js
// 0. Abstract Character 
class Character {

  constructor(charType, charName, charWeapon) {

    this.charType = charType;
    this.charName = charName;
    this.charWeapon = charWeapon;
  }
}

// 1. Concrete Spy
class Spy extends Character {

  constructor(...args) {

    super(...args);
    this.catchPhrase = `The name is ${this.charName}, ${this.charName}`;
  }
}

// 2. Concrete Programmer
class Programmer extends Character {

  constructor(...args) {

    super(...args);
    this.catchPhrase = 'I\'m invincible!';
  }
}

// 3. Concrete Engineer 
class Engineer extends Character {

  constructor(...args) {

    super(...args);
    this.catchPhrase = 'Grow up, Bond.';
  }
}

File: Python/factory_demo/concrete.py
# --- Abstract Character ---
class Character:

    def __init__(self, char_type, char_name, char_weapon):

        self.char_type = char_type
        self.char_name = char_name
        self.char_weapon = char_weapon

    def char_data(self):

        # display all character data
        print(f'{self.char_type.upper()}: {self.char_name} uses a {self.char_weapon}')


# --- Concrete Spy ---
class Spy(Character):

    def __init__(self, char_type, char_name, char_weapon):

        super().__init__(char_type, char_name, char_weapon)
        self.catch_phrase = f'My name is {self.char_name}, {self.char_name}.'


# --- Concrete Programmer ---
class Programmer(Character):

    def __init__(self, char_type, char_name, char_weapon):

        super().__init__(char_type, char_name, char_weapon)
        self.catch_phrase = 'I\'m invincible!'


# --- Concrete Engineer ---
class Engineer(Character):

    def __init__(self, char_type, char_name, char_weapon):

        super().__init__(char_type, char_name, char_weapon)
        self.catch_phrase = 'Grow up, Bond.'
